### PR TITLE
Addstories

### PR DESCRIPTION
**Pull request by MrButak**

### CODE REVIEW

File: controllers/stories.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+const storyManager = require('../public/javascripts/data/storyManager');
+const paragraph = require('../public/javascripts/data/paragraph');
+
+exports.getAllStories = function(req, res, next) {
+
+    let currentUserName = res.locals.userName;
```
Comment (JohnnyCurran) [inline controllers/stories.js:6]: Prefer `req.session` here

File: controllers/stories.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    let storyTitle = req.body.addStoryInput;
+    let userId = req.session.user['id'];
+    
+    storyManager.addStory(storyTitle, userId);
+    res.redirect('/');
+};
+
+exports.viewStory = function(req, res, next) {
+
+    let currentUserName = res.locals.userName;
```
Comment (JohnnyCurran) [inline controllers/stories.js:23]: Prefer `req.session` for username retrieval

File: controllers/stories.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    res.render('story', { story: storyManager.getStory(storyId), currentUserName: currentUserName });
+};
+
+// POST request /addparagraph from form on /addparagraph
+exports.addParagraph = function(req, res, next) {
+    
+    // Do a check here to see if there is a paragraph to submit
+    // This /story POST is shared between the button on the home page to view a story and submitting a paragraph on /story
+    if(!req.body.paragraph_input) {
+        next();
```
Comment (JohnnyCurran) [inline controllers/stories.js:35]: Is this a validation check? Do you need to return some view data indicating to the user that the paragraph input form was empty and that is why their paragraph was not added?

File: public/javascripts/data/storyManager.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
+const Database = require('better-sqlite3');
+
+exports.displayAllStories = () => {
```
Comment (JohnnyCurran) [inline public/javascripts/data/storyManager.js:3]: You might consider accepting a limit parameter here in `displayAllStories` which would limit the number of story records returned by the DB. The reason being as your application gets more and more story submissions, the home page will continue to grow and become extremely long. When you want to render the homepage, you could pass a limit parameter of 5 or 10 and default to all stories if there's no limit parameter passed.

It's not a hard requirement, just a design consideration. This is the part of application development that begins to bleed over to more of the soft requirements side

File: public/javascripts/data/storyManager.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    
+    // Get all stories (an array of objects)
+    let db = new Database('public/javascripts/data/stories.db');
+    let allStories = db.prepare('SELECT * FROM stories').all();
```
Comment (JohnnyCurran) [inline public/javascripts/data/storyManager.js:7]: Consider using an `ORDER BY` clause to obtain the most recent stories inserted to the database

File: public/javascripts/data/storyManager.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+   
+    db.close();
+    return(allStories);
+};
+
+// Function writes a new story (title) to the database
+exports.addStory = (storyTitle, userId) => {
+
+    let db = new Database('public/javascripts/data/stories.db');
+    const new_user = db.prepare('INSERT INTO stories (user_id, title) VALUES (?, ?)').run(userId, storyTitle);
```
Comment (JohnnyCurran) [inline public/javascripts/data/storyManager.js:30]: I'd rename this variable to be `new_story` or simply `story`. Also, prefer `let` over `const`

File: public/javascripts/data/storyManager.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    db.close();
+    return;
```
Comment (JohnnyCurran) [inline public/javascripts/data/storyManager.js:32]: Do you want to return the story from the insert? Consider adding a `RETURNING` clause to your `INSERT` statement

https://www.sqlite.org/lang_returning.html

File: public/javascripts/data/storyManager.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+};
+
+// Function gets 1 story and all of it's paragraphs
+exports.getStory = (storyId) => {
+    
+    let db = new Database('public/javascripts/data/stories.db');
+    let story = db.prepare('SELECT * FROM stories WHERE id = (?)').get(storyId);
+    // Add the paragraphs as a property
+    story['paragraphs'] = db.prepare('SELECT content FROM paragraphs WHERE stories_id = (?)').all(storyId);
```
Comment (JohnnyCurran) [inline public/javascripts/data/storyManager.js:41]: You may want to select both content and ID from the paragraphs. It's generally helpful to have as a property hanging around, even if your application doesn't require it in its current state

File: views/index.ejs [hunk 9]
```diff
@@ -1,16 +1,38 @@
...
-<p>Current user</p>
-<p><%= currentUserName %></p>
-<a href="/logout">logout</a>
-<ul>
-    <% paragraphs.forEach(function(paragraph) { %>
-        <li><%= paragraph['content'] %></li>
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <%- include('../views/partials/head.ejs'); %>
```
Comment (JohnnyCurran) [inline views/index.ejs:4]: Nice use of partials for reusing elements

File: views/index.ejs [hunk 10]
```diff
@@ -1,16 +1,38 @@
...
+
+    <h1>I'm the stories app</h1>
+    <p>Current user: <%= currentUserName %></p>
+    <a href="/logout">Logout</a>
+
+    <!-- This loops through the stories array, displaying each title and the first 5 
+        paragraphs, and a hidden field with the story id -->
+
+    <div class="storyTitlesWrapper">
+        <% for (var i = stories.length - 1; i > -1;  i-- ) { %>
```
Comment (JohnnyCurran) [inline views/index.ejs:17]: Consider using a `forEach` loop here

File: views/story.ejs [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    <%- include('../views/partials/head.ejs'); %>
+</head>
+
+<body>
+    <h2>Welcome to the story page</h2>
+    <p>Current user: <%= currentUserName %></p>
+    <a href="/logout">Logout</a>
+
+    <h2><%= story['title'] %></h2>
+    <% for (var i = 0; i < story['paragraphs'].length;  i++ ) { %>
```
Comment (JohnnyCurran) [inline views/story.ejs:13]: Consider using a `forEach` loop here

**Review by JohnnyCurran — COMMENTED** [review]
The app is coming along real nicely. Just about to MVP!

File: controllers/stories.js (hunk 3 above)
Comment (MrButak) [reply controllers/stories.js:35]: Here I'm sharing two POST requests on  /stories routing. They both have the same middleware  except this one function
` stories.addParagraph`. So it's just a quick check to see if I can bypass that function (if they're on the home page and clicking the button to view the story (/story POST)

This is the route each of the two POST requests get sent to
```
// view story button POST request on "/ homepage" and submit paragraph POST request on "/story"
router.post('/story', handleAuth.requireLogin, handleAuth.currentUser, stories.addParagraph, stories.viewStory);
```

Comment (JohnnyCurran) [reply controllers/stories.js:35]: I gotcha. For your view story route, `/story`, I'd recommend using a GET. You can pass the Id to the router and with your link. For instance:

index.ejs:
```html
<a href="/stories/<%= story.id %>"
```

router.js
```js
router.get('/story/:id', handleAuth.requireLogin, handleAuth.currentUser, stories.viewStory)
```

File: public/javascripts/data/storyManager.js (hunk 5 above)
Comment (MrButak) [reply public/javascripts/data/storyManager.js:7]: I was able to do this and use the forEach loop. I think it is much better.

`// Get all stories (an array of objects)
    let db = new Database('public/javascripts/data/stories.db');
    let allStories = db.prepare('SELECT * FROM stories ORDER BY id DESC').all();

    // Get the first five paragraphs (if that many) from each story.
    let i = allStories.length;
    allStories.forEach( (story) => {
        let paraStmt = db.prepare('SELECT * from paragraphs WHERE stories_id = (?) LIMIT 5');
        story['paragraphs'] = paraStmt.all(i);
        i--;
    });
`

Comment (JohnnyCurran) [reply public/javascripts/data/storyManager.js:7]: You won't need the `i` variable - Use the `id` property of the `story` object

File: controllers/stories.js (hunk 3 above)
Comment (MrButak) [reply controllers/stories.js:35]: Great advice. I was able to implement this.

`exports.viewStory = function(req, res, next) {
    
    let currentUserName = req.session.user['user_name'];
    
    // GET request
    if(req.method == "GET") {

       let storyId =  req.query['storyId'];
       res.render(`story`, { story: storyManager.getStory(storyId), currentUserName: currentUserName });
    }
    // POST request
    else {
        
        let storyId = req.body['storyId'];
        res.render('story', { story: storyManager.getStory(storyId), currentUserName: currentUserName });
    };
};`

File: public/javascripts/data/storyManager.js (hunk 7 above)
Comment (MrButak) [reply public/javascripts/data/storyManager.js:32]: Here after inserting the story into the database, I'm redirecting back to the home page which displays all of the stories.

`router.get('/', handleAuth.requireLogin, handleAuth.currentUser, stories.getAllStories);`

```
exports.addStory = function(req, res, next) {

    let storyTitle = req.body.addStoryInput;
    let userId = req.session.user['id'];
    
    storyManager.addStory(storyTitle, userId);
    res.redirect('/');
};
```

**Pull request merged**